lib/sla.js

```diff
@@ -231,101 +305,251 @@ function originateChannel(client, inbound, bridge, trunkData) {
+ * @param {Object} outbound - the outbound channel object
+ * @param {Object} bridge - the bridge used in this outbound call
+ */
+function originateOutbound(client, trunkData, outbound, bridge) {
+  var toDial = '';
+  var callInProgress = false;
+  var channelToDial;
+  var defer = Q.defer();
 
-  trunkData.getAllStations()
-    .then(function (stations) {
-      if (stations.length) {
-        stations.forEach(function(station) {
-          dialed.push({endpoint: station.endpoint, channel: client.Channel()});
-        });
-
-        dialed.forEach(function (object) {
-          var originate = Q.denodeify(object.channel.originate.bind(
-              object.channel));
-          originate({endpoint: object.endpoint, app: 'sla',
-              appArgs: 'dialed', timeout: 10})
-            .catch(function (err) {
-              defer.reject(err);
-            });
+  getState(client, bridge.name).then(function (currentState) {
+    if (currentState === states.INUSE || currentState === states.RINGING ||
+      currentState === states.BUSY) {
+      callInProgress = true;
+    }
 
-          object.channel.once('StasisStart', dialedEnteredStasis);
-          object.channel.once('ChannelDestroyed', dialedChannelHungup);
-        });
+    if (!trunkData[0]) {
+      defer.reject(new customError.CustomError('NoTrunks', 'There are no ' +
+         'trunks with which to outbound dial'));
+    } else if (!callInProgress) {
+      var playback = client.Playback();
+      outbound.play({media: 'sound:pls-entr-num-uwish2-call'}, playback);
 
-        updateState(client, bridge.name, ringing);
-        inbound.once('ChannelHangupRequest', inboundHungup);
-      }
-      else {
-        defer.reject(new customError.CustomError('NoStations',
-              'No stations in this trunk.'));
-      }
-    });
+      updateState(client, bridge.name, states.BUSY);
+      outbound.on('ChannelDtmfReceived', dtmfReceived);
+      outbound.once('ChannelHangupRequest', outboundHungupEarly);
+    } else {
+      console.log('CALL IN PROGRESS');
+      defer.resolve();
+    }
+  });
   return defer.promise;
+  /**
+   * The function that gets called once the dialed channel enters Stasis.
+   *   Makes the extension busy and clears eventListeners
+   * @param {Object} event - the information related to the StasisStart event
+   * @param {Object} line - the line that has entered Stasis
+   */
+  function dialedEnteredStasis(event, line) {
+    updateState(client, bridge.name, states.INUSE);
+    outbound.removeListener('ChannelHangupRequest', outboundHungup);
+    outbound.removeListener('ChannelDtmfReceived', dtmfReceived);
+    var answer = Q.denodeify(line.answer.bind(line));
+    answer();
+    defer.resolve(line);
+  }
+
+  /**
+   * Function that gets called when a dialed channel hangs up
+   * @param {Object} event - the event object related to this hang up
+   */
+  function dialedChannelHungup(event) {
```

Comment:
Ashley also mentioned renaming the event handlers to make it more clear that they are event handlers. This was also not addressed. Although here I would suggest not using on_event_name and instead using onEventName to match the camel case used throughout the project.